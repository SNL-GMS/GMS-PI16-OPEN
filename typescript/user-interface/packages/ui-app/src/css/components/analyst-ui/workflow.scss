@import '../../gms-colors.scss';
@import '../../gms-fonts.scss';

$BLOCK_HEIGHT_PX: 24px;
$ROW_COMPUTED_HEIGHT_PX: 26px;
$BLOCK_WIDTH_PX: 140px;
$TABLE_MARGIN: 34px;
$DIVIDER_MARGIN: 31px;
//The table margin plus some manual fiddling
$WIDTH_OF_LEFT_CURTAIN: 43px;
$gap-between-rows: 0.75em;

.gms-workflow-time-axis {
  path,
  line {
    fill: none;
    stroke: $gms-prominent;
    shape-rendering: crispEdges;
  }
  text {
    font-family: $gms-sans;
    font-size: 12px;
    fill: $gms-main;
    &.day-label {
      stroke: $workflow-day-divider;
      font-size: 13px;
    }
  }
}

.workflow-display-window {
  height: 100%;
}

.workflow-panel {
  height: 100%;
  width: 100%;
}

.workflow-scroll-wrapper {
  max-height: calc(100% - 55px);
  overflow-y: scroll;
}

.interval-table-wrapper {
  flex: 1 0 auto;
  display: flex;
  margin-left: $TABLE_MARGIN;
  margin-right: $BLOCK_WIDTH_PX;
  flex-direction: column;
  overflow-x: scroll;
  overflow-y: hidden;
  position: static;
  gap: 2px;
}

.interval-table-wrapper-wrapper {
  position: relative;
  overflow-y: hidden;
  overflow-x: hidden;
}

// Left is set to account for the workflow-table-wrapper's left margin, minus the size of the button
.stage-row__expand-button {
  height: $BLOCK_HEIGHT_PX;
  border: 1px solid $gms-recessed;
  width: $BLOCK_HEIGHT_PX;
  transition: margin-bottom 0.15s;
  min-height: $ROW_COMPUTED_HEIGHT_PX;
  min-width: $ROW_COMPUTED_HEIGHT_PX;
  padding: 0;
  position: absolute;
  left: 7px;
  margin-top: -1px;
  z-index: 5;
}

.stage-row {
  display: flex;
  gap: 2px;
}

.stage-column {
  display: flex;
  flex-direction: column;
  gap: 2px;
}

.internal-row {
  display: grid;
  grid-auto-columns: 8px;
  gap: 2px;
}

.workflow-tooltip-container {
  display: grid;
  grid-template-columns: 2fr 3fr;
  margin: 0 auto;
  gap: 0.25em;
  padding-bottom: 0.5em;
  border-bottom: 1px dashed $gms-soft;
}

.workflow-tooltip-container .label-value__label {
  font-weight: bold;
}

.workflow-tooltip-container:last-child {
  border-bottom: none;
  padding-bottom: 0;
}

.workflow-tooltip-container + .workflow-tooltip-container {
  padding-top: 0.5em;
}

// The absolute position lets it float to the left over the workflow display
// The margin right and top are for alignment
.workflow-table-label {
  height: $ROW_COMPUTED_HEIGHT_PX;
  width: $BLOCK_WIDTH_PX;
  border: 1px solid;
  display: flex;
  border-color: $gms-background;
  justify-content: center;
  align-items: center;
  background: $gms-soft;
  flex: 0 0 auto;
  position: absolute;
  right: 0px;
  margin-right: 4px;
  margin-top: -1px;
  z-index: 5;
}

.workflow-table-label--activity {
  filter: $interval-activity-darken-cell-by;
}

.workflow-row {
  position: relative;
  display: flex;
  margin-bottom: 2px;
}

.interval-cell {
  display: flex;
  flex-direction: column;
  height: 100%;
  height: $BLOCK_HEIGHT_PX;
  border: 1px solid $gms-background;
  align-items: center;
  justify-content: center;
  margin: 0;
  gap: 2px;
  overflow: hidden;
  white-space: nowrap;
  cursor: pointer;
  .percent-bar {
    height: 100%;
    width: 100%;
  }
}

.interval-cell:hover {
  filter: brightness(1.2);
}

.interval-cell--in-progress {
  background-color: $interval-in-progress-fill;
  border: 1px solid $interval-in-progress-border;
  .percent-bar {
    background-color: scale-color(
      $color: $interval-in-progress-fill,
      $lightness: 25%,
      $saturation: -15%
    );
  }
}

.interval-cell--complete {
  background-color: $interval-complete-fill;
  border: 1px solid $interval-complete-border;
}

.interval-cell--not-complete {
  background-color: $interval-not-complete-fill;
  border: 1px solid $interval-not-complete-border;
}

.interval-cell--not-started {
  background-color: $interval-not-started-fill;
  border: 1px solid $interval-not-started-border;
}

.interval-cell--skipped {
  background-color: $interval-skipped-fill;
  border: 1px solid $interval-skipped-border;
}

.interval-cell--failed {
  background-color: $interval-failed-fill;
  border: 1px solid $interval-failed-border;
}

.interval-cell--selected {
  border: 2px solid $gms-selection;
}

.interval-cell--activity-cell {
  filter: $interval-activity-darken-cell-by;
}

.interval-cell--activity-cell:hover {
  filter: brightness(1);
}

.interval-cell--stale {
  filter: brightness(0.5);
}

.workflow-day-divider {
  position: absolute;
  top: 0px;
  bottom: 0px;
  width: 4px;
  z-index: 1;
  margin-left: $DIVIDER_MARGIN;
}

// To account for the offset of the workflow-table, the time axis needs to be moved to left by 1
.time-axis-wrapper {
  margin-left: -1px;
  width: calc(100%-131px);
  height: 25px;
  flex: 0 0 auto;
  position: static;
  z-index: 1;
}

.time-axis {
  width: 100%;
  height: 25px;
  overflow: hidden;
}

.interval-table__curtain-left {
  width: $WIDTH_OF_LEFT_CURTAIN;
  position: absolute;
  background-color: $gms-background;
  height: 100vh;
  z-index: 2;
  left: -6px;
}

.workflow-ellipsis {
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
  width: 100px;
  text-align: center;
}

.workflow-open-anything-row {
  display: flex;
  flex-flow: wrap;
  box-sizing: border-box;
  align-items: center;
  min-width: 56ch;
  & > * {
    margin-left: 0.5rem;
  }
}

.open-anything-contents .dropdown-container {
  display: flex;
  justify-content: space-between;
}

.workflow-open-anything-row .dropdown-container .span {
  display: flex;
}

.open-anything-contents .date-range-picker {
  width: 100%;
  display: flex;
  margin-bottom: 5px;
}

.open-anything-contents .date-range-picker .date-range-picker__range-input {
  width: 65%;
  margin: 0px;
}

.open-anything-contents
  .date-range-picker
  .date-range-picker__range-input
  .bp3-popover-target
  .bp3-control-group
  * {
  flex-shrink: 1;
}

.open-anything-contents .date-range-picker .date-range-picker__label {
  font-size: 1.125em;
  font-weight: 600;
  margin: 0px;
  width: 35%;
}

.open-anything-contents .dropdown-label {
  font-size: 1.125em;
  font-weight: 600;
  margin: 0px;
  width: 35%;
}

.open-anything-contents .dropdown-selector {
  margin-bottom: 5px;
  width: 65%;
}

.interval-confirmation-contents {
  display: flex;
}

.interval-confirmation-text {
  width: 90%;
}

.interval-confirmation-icon {
  width: 10%;
}

.interval-confirmation-header {
  font-weight: bold;
}

.open-anything-error {
  color: $gms-strong-warning;
  display: flex;
  min-height: 20px;
}
.open-anything-error-text {
  margin-left: 5px;
  width: calc(100%-20px);
}
