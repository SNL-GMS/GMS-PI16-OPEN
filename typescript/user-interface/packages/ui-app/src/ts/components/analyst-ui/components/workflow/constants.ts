import { SECONDS_IN_HOUR } from '@gms/common-util';

export const PIXELS_PER_HOUR = 120;
export const PIXELS_PER_SECOND = PIXELS_PER_HOUR / SECONDS_IN_HOUR;
