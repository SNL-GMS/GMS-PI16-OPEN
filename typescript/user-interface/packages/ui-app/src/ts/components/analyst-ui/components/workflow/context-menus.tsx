import { Menu, MenuItem } from '@blueprintjs/core';
import { AppState } from '@gms/ui-state';
import React from 'react';
import { useSelector } from 'react-redux';

import { IntervalContextMenuProps } from './types';

/**
 * Row entry wrapper for ActivityInterval cells
 */
export const IntervalContextMenu: React.FunctionComponent<IntervalContextMenuProps> = (
  props: IntervalContextMenuProps
) => {
  const workflowState = useSelector((state: AppState) => state.analystWorkspaceState.workflowState);
  const {
    interval,
    isSelectedInterval,
    allActivitiesOpenForSelectedInterval,
    openCallback,
    closeCallback
  } = props;

  const isDisabled =
    (allActivitiesOpenForSelectedInterval &&
      isSelectedInterval &&
      workflowState.openIntervalName === interval.name) ||
    (isSelectedInterval && interval.name !== workflowState.openIntervalName);
  return (
    <Menu>
      <MenuItem
        className="menu-item-open-interval"
        text="Open Interval"
        disabled={isDisabled}
        onClick={() => openCallback(interval)}
      />
      <MenuItem
        className="menu-item-close-interval"
        text="Close Interval"
        disabled={!isSelectedInterval}
        onClick={() => closeCallback(interval)}
      />
    </Menu>
  );
};
