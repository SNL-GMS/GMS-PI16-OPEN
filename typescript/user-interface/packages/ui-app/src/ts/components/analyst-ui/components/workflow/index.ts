export { WorkflowComponent as Workflow } from './workflow-component';
