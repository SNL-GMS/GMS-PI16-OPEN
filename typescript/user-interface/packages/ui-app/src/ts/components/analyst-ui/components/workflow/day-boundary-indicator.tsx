import { CommonTypes } from '@gms/common-model';
import { diff, MILLISECONDS_IN_SECOND } from '@gms/common-util';
import * as d3 from 'd3';
import isEqual from 'lodash/isEqual';
import moment from 'moment';
import React from 'react';

import { PIXELS_PER_SECOND } from './constants';

/**
 * DayBoundaryIndicator Props
 */
export interface DayBoundaryIndicatorProps {
  timeRange: CommonTypes.TimeRange;
}

/**
 * @DayBoundaryIndicator
 * The marker between days in the workflow table
 * It is kept in synch with the workflow table through its onScroll callback
 * Line indicating the transition from one day from the next
 */

export class DayBoundaryIndicator extends React.Component<DayBoundaryIndicatorProps> {
  /** Day indicators element ref array */
  private dayIndicators: HTMLDivElement[] = [];

  public shouldComponentUpdate(nextProps: DayBoundaryIndicatorProps): boolean {
    const { timeRange } = this.props;
    return !isEqual(nextProps.timeRange, timeRange);
  }

  // eslint-disable-next-line react/sort-comp
  public render(): JSX.Element[] {
    const { timeRange } = this.props;

    // add a slight padding to ensure that day boundary is displayed and rendered on the edges
    const dayBoundaryPaddingSecs = 5;

    const totalDurationSecs = diff(timeRange?.startTimeSecs, timeRange?.endTimeSecs);
    const rowWidthPx = Math.ceil(totalDurationSecs * PIXELS_PER_SECOND);

    this.dayIndicators = [];

    const dividers = d3.utcDay
      .every(1)
      .range(
        moment.unix(timeRange?.startTimeSecs - dayBoundaryPaddingSecs).toDate(),
        moment.unix(timeRange?.endTimeSecs + dayBoundaryPaddingSecs).toDate()
      );
    return dividers.map(date => {
      const widthPx = 4;
      // calculation that sets the initial offset of the day indicator
      const left = `calc(${
        ((date.valueOf() / MILLISECONDS_IN_SECOND - timeRange?.startTimeSecs) /
          (timeRange?.endTimeSecs - timeRange?.startTimeSecs)) *
        rowWidthPx
      }px + ${widthPx / 4}px)`;
      // The divider's base offset from the workflow is set to the 'left' style - it is based
      // on the timeSecs of the day boundary and does not change.

      return (
        <div
          key={date.valueOf()}
          className="workflow-day-divider"
          style={{
            left,
            backgroundColor: '#ffc940' // asked to add the color this way to be able to override with configuration in future
          }}
          ref={ref => {
            this.dayIndicators.push(ref);
          }}
        />
      );
    });
  }

  /**
   * Synchronizes the day indicator with the main display
   *
   * @param scrollTo pixel value reflecting how scrolled the workflow is
   */
  public readonly scrollDayIndicator = (scrollTo: number): void => {
    // hard-coded value to account for distance between the workflow table and the edge of the workflow display
    const marginOfIntervalTable = 31;
    // removes any references that have become invalid (which happens every rerender)
    this.dayIndicators = this.dayIndicators.filter(di => di !== undefined && di !== null);
    // To enable scrolling, we change the marginLeft of the interval boundary. The 'left' position property
    // Remains static
    this.dayIndicators.forEach(di => {
      // eslint-disable-next-line no-param-reassign
      di.style.marginLeft = `${(-scrollTo + marginOfIntervalTable).toString()}px`;
    });
  };
}
