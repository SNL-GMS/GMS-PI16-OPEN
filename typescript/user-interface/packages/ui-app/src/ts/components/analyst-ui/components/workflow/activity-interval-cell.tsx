import { ContextMenu } from '@blueprintjs/core';
import { WorkflowTypes } from '@gms/common-model';
import { AppState } from '@gms/ui-state';
import classNames from 'classnames';
import includes from 'lodash/includes';
import React from 'react';
import { Provider, useSelector, useStore } from 'react-redux';

import { IntervalContextMenu } from './context-menus';
import { ActivityIntervalCellProps } from './types';
import { WorkflowContext } from './workflow-context';
import { determineCellSize } from './workflow-util';

/**
 * Determines the text for the cell based on status and activeAnalyst list
 * TODO for complete name needs to be analyst that marked complete
 *
 * @param status interval status
 * @param activeAnalysts list of activeAnalyst
 * @returns text for cell to display
 */
export const determineTextForCell = (
  status: WorkflowTypes.IntervalStatus,
  activeAnalysts: string[]
): string => {
  if (!activeAnalysts || activeAnalysts.length === 0) return '';

  switch (status) {
    case WorkflowTypes.IntervalStatus.NOT_STARTED:
    case WorkflowTypes.IntervalStatus.NOT_COMPLETE:
      return '';
    case WorkflowTypes.IntervalStatus.COMPLETE:
      return activeAnalysts[0];
    default:
      return `${activeAnalysts[0]}${
        activeAnalysts.length > 1 ? ` + ${activeAnalysts.length - 1}` : ''
      }`;
  }
};

export const preventDefaultEvent = (event: React.MouseEvent<HTMLDivElement, MouseEvent>): void => {
  event.preventDefault();
};

export const ActivityIntervalCell: React.FunctionComponent<ActivityIntervalCellProps> = React.memo(
  (props: ActivityIntervalCellProps) => {
    const { activityInterval } = props;
    const workflowState = useSelector(
      (state: AppState) => state.analystWorkspaceState.workflowState
    );
    const context = React.useContext(WorkflowContext);
    const isSelected =
      workflowState.openIntervalName === activityInterval.stageName &&
      includes(workflowState.openActivityNames, activityInterval.name) &&
      workflowState.timeRange?.startTimeSecs === activityInterval.startTime;
    const isStale = context.staleStartTime > activityInterval.startTime;
    const cellClass = classNames({
      'interval-cell': true,
      'interval-cell--selected': isSelected,
      'interval-cell--not-complete':
        activityInterval.status === WorkflowTypes.IntervalStatus.NOT_COMPLETE,
      'interval-cell--in-progress':
        activityInterval.status === WorkflowTypes.IntervalStatus.IN_PROGRESS,
      'interval-cell--not-started':
        activityInterval.status === WorkflowTypes.IntervalStatus.NOT_STARTED,
      'interval-cell--complete': activityInterval.status === WorkflowTypes.IntervalStatus.COMPLETE,
      'interval-cell--activity-cell': true,
      'interval-cell--stale': isStale
    });
    const store = useStore();
    return (
      <div
        key={activityInterval.startTime}
        data-cy={`${activityInterval.startTime}-${activityInterval.name}`}
        className={cellClass}
        style={{
          gridColumn: `span ${determineCellSize(
            activityInterval.startTime,
            activityInterval.endTime
          )}`
        }}
        onContextMenu={
          !isStale
            ? e => {
                e.preventDefault();
                ContextMenu.show(
                  <Provider store={store}>
                    <IntervalContextMenu
                      interval={activityInterval}
                      isSelectedInterval={isSelected}
                      allActivitiesOpenForSelectedInterval={
                        context.allActivitiesOpenForSelectedInterval
                      }
                      openCallback={context.openConfirmationPrompt}
                      closeCallback={context.closeConfirmationPrompt}
                    />
                  </Provider>,
                  {
                    left: e.clientX,
                    top: e.clientY
                  },
                  undefined,
                  true
                );
              }
            : preventDefaultEvent
        }
        title={determineTextForCell(activityInterval.status, activityInterval.activeAnalysts)}
      >
        <span className="workflow-ellipsis">
          {determineTextForCell(activityInterval.status, activityInterval.activeAnalysts)}
        </span>
      </div>
    );
  }
);
