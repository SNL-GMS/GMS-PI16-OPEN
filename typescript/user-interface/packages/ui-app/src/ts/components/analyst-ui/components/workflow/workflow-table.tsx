import { WorkflowTypes } from '@gms/common-model';
import flatMap from 'lodash/flatMap';
import React from 'react';

import { DayBoundaryIndicator } from './day-boundary-indicator';
import { WorkflowTableProps } from './types';
import { WorkflowRow } from './workflow-row';

/**
 * Time axis for the Workflow display
 */
export class WorkflowTable extends React.PureComponent<WorkflowTableProps> {
  public intervalTableWrapper: HTMLDivElement;

  public dayBoundaryIndicator: DayBoundaryIndicator;

  public render(): JSX.Element {
    const { timeRange, stageIntervals, workflow, onScroll } = this.props;

    return (
      <div className="workflow-scroll-wrapper">
        <div className="interval-table__curtain-left" />
        <div className="interval-table-wrapper-wrapper">
          <DayBoundaryIndicator
            timeRange={timeRange}
            ref={ref => {
              this.dayBoundaryIndicator = ref;
            }}
          />
          <div className="interval-table-scroll-wrapper" />
          <div
            className="interval-table-wrapper"
            ref={ref => {
              this.intervalTableWrapper = ref;
            }}
            onScroll={event => onScroll(event)}
          >
            {flatMap(
              workflow.stages.map(stage => {
                let subRowNames = [];
                if (stage.mode === WorkflowTypes.StageMode.AUTOMATIC) {
                  const autoStage = stage as WorkflowTypes.AutomaticProcessingStage;
                  subRowNames = autoStage.sequences.map(sequence => sequence.name);
                } else {
                  const interactiveStage = stage as WorkflowTypes.InteractiveAnalysisStage;
                  subRowNames = interactiveStage.activities.map(activity => activity.name);
                }
                return (
                  <WorkflowRow
                    key={stage.name}
                    subRowNames={subRowNames}
                    stageIntervals={stageIntervals.get(stage.name)}
                  />
                );
              })
            )}
          </div>
          {/*
              The curtain is set s.t. it prevents the day boundary indicator
              from displaying off the edges of the workflow table
            */}
        </div>
      </div>
    );
  }
}
