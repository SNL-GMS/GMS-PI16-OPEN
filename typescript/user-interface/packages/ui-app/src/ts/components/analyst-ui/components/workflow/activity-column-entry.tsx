import React from 'react';

import { ActivityIntervalCell } from './activity-interval-cell';
import { ActivityColumnEntryProps } from './types';

/**
 * Row entry wrapper for ActivityInterval cells
 */
// eslint-disable-next-line react/display-name
export const ActivityColumnEntry: React.FunctionComponent<ActivityColumnEntryProps> = React.memo(
  (props: ActivityColumnEntryProps) => {
    const { activityInterval } = props;
    return <ActivityIntervalCell activityInterval={activityInterval} key={activityInterval.name} />;
  }
);
