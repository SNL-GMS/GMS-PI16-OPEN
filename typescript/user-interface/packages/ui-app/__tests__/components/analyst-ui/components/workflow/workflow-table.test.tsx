import Immutable from 'immutable';
import React from 'react';

import { WorkflowTable } from '../../../../../src/ts/components/analyst-ui/components/workflow/workflow-table';
import { workflow } from '../../../../__data__/workflow-data';

window.ResizeObserver = jest.fn(() => {
  return { observe: jest.fn(), disconnect: jest.fn(), unobserve: jest.fn() };
});
const globalAny: any = global;
globalAny.ResizeObserver = window.ResizeObserver;
globalAny.DOMRect = jest.fn(() => ({}));

// eslint-disable-next-line @typescript-eslint/no-var-requires, @typescript-eslint/no-require-imports
const Enzyme = require('enzyme');

// set up window alert and open so we don't see errors
window.alert = jest.fn();
window.open = jest.fn();

const MOCK_TIME = 1606818240000;
global.Date.now = jest.fn(() => MOCK_TIME);

describe('Workflow Table', () => {
  it('is exported', () => {
    expect(WorkflowTable).toBeDefined();
  });

  it('matches snapshot', () => {
    const component = Enzyme.mount(
      <WorkflowTable
        timeRange={{
          startTimeSecs: 0,
          endTimeSecs: 360000
        }}
        stageIntervals={Immutable.Map()}
        workflow={workflow}
        onScroll={jest.fn()}
      />
    );
    expect(component).toMatchSnapshot();
  });
});
