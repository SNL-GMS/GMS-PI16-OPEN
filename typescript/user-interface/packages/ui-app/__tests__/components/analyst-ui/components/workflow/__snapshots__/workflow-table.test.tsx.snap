// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Workflow Table matches snapshot 1`] = `
<WorkflowTable
  onScroll={[MockFunction]}
  stageIntervals={Immutable.Map {}}
  timeRange={
    Object {
      "endTimeSecs": 360000,
      "startTimeSecs": 0,
    }
  }
  workflow={
    Object {
      "name": "workflow",
      "stages": Array [
        Object {
          "activities": Array [
            Object {
              "analysisMode": "EVENT_REVIEW",
              "name": "name",
              "stationGroup": Object {
                "description": "Hi",
                "effectiveAt": 1000,
                "name": "ALL_1",
              },
            },
          ],
          "duration": 50,
          "mode": "INTERACTIVE",
          "name": "name",
        },
        Object {
          "duration": 5000,
          "mode": "AUTOMATIC",
          "name": "name",
          "sequences": Array [
            Object {
              "name": "name",
              "steps": Array [],
            },
          ],
        },
      ],
    }
  }
>
  <div
    className="workflow-scroll-wrapper"
  >
    <div
      className="interval-table__curtain-left"
    />
    <div
      className="interval-table-wrapper-wrapper"
    >
      <DayBoundaryIndicator
        timeRange={
          Object {
            "endTimeSecs": 360000,
            "startTimeSecs": 0,
          }
        }
      >
        <div
          className="workflow-day-divider"
          key="0"
          style={
            Object {
              "backgroundColor": "#ffc940",
              "left": "calc(0px + 1px)",
            }
          }
        />
        <div
          className="workflow-day-divider"
          key="86400000"
          style={
            Object {
              "backgroundColor": "#ffc940",
              "left": "calc(2880px + 1px)",
            }
          }
        />
        <div
          className="workflow-day-divider"
          key="172800000"
          style={
            Object {
              "backgroundColor": "#ffc940",
              "left": "calc(5760px + 1px)",
            }
          }
        />
        <div
          className="workflow-day-divider"
          key="259200000"
          style={
            Object {
              "backgroundColor": "#ffc940",
              "left": "calc(8640px + 1px)",
            }
          }
        />
        <div
          className="workflow-day-divider"
          key="345600000"
          style={
            Object {
              "backgroundColor": "#ffc940",
              "left": "calc(11520px + 1px)",
            }
          }
        />
      </DayBoundaryIndicator>
      <div
        className="interval-table-scroll-wrapper"
      />
      <div
        className="interval-table-wrapper"
        onScroll={[Function]}
      >
        <WorkflowRow
          key="name"
          subRowNames={
            Array [
              "name",
            ]
          }
        >
          <div />
        </WorkflowRow>
        <WorkflowRow
          key="name"
          subRowNames={
            Array [
              "name",
            ]
          }
        >
          <div />
        </WorkflowRow>
      </div>
    </div>
  </div>
</WorkflowTable>
`;
